### ISSUE TITLE
Hangs forever with 780000 samples and 6 variables

### ISSUE THREAD

**Issue by slizovskiy** [bgrimstad/splinter#49]
Hi!  I have 6 variables on a regular grid,  and constructing of even the LINEAR spline loads the processor forever.  Is that a bug?

Thanks,
Sergey

**Comment by slizovskiy** [comment]
I may post a bit of my code as I really need help with this issue:  
  First part constructs my data table:
  DenseVector x(6);
  double y;
  //double mu=0.01;
  //double T=0.05;
  //Notations:  x(0) is ln omega  , x(1) = radial momentum,  x(2) = angle , x(3) = kz , x(4) = T , x(5) = mu
  double lnomegamin=-6;
  double lnomegamax=3;

  int kzmax=10;
  int kphimax=8;
  int krmax=25;
  int omegamax=13;
  double mumin=-0.05;
  double mumax=0.05;
  int imumax=6;

  double Tmin=0.001;
  double Tmax=0.05;
  int iTmax=5;
  bool ComputeTable = false;
  std::string filename("/home/sergey/Article/Joseph/Lifshitz/suscept.dat");
  DataTable *samples;

  if (ComputeTable)
  { samples= new DataTable();
    for(int imu=0;imu<imumax;imu++){
      double mu=mumin+(imu)*(mumax-mumin)/(imumax-1);
      x(5)=mu;
      for(int iT=0;iT<iTmax;iT++){
          double T=Tmin+(iT)*(Tmax-Tmin)/(iTmax-1);
          x(4)=T;
          for(int i0 = 0; i0 < omegamax ; i0++)
          { x(0)=lnomegamin+(i0)*(lnomegamax-lnomegamin)/(omegamax-1);
              cout<<"ln(omega)="<<x(0)<<" ;  ";
              x(1)=0.0;
              for(int i1 = 0; i1 < krmax; i1++) {
                  x(2)=0;
                  for(int i2 = 0; i2 < kphimax; i2++){
                      x(3)=0;
                      for(int i3 = 0; i3 < kzmax; i3++){
                          // Sample function at x
                          double q[3]={x(1)*cos(x(2)), x(1)*sin(x(2)), x(3)};
                          y = Chi0(q,exp(x(0)),mu,T,40000);
//                          cout<<"q="<<q[0]<<" "<<q[1]<<" "<<q[2];
                          // Store sample
                          samples->addSample(x,y);
  //                        cout<<y<<" ";
                          x(3)+=M_PI/kzmax;
                      }
                      x(2)+= M_PI/(3*kphimax);
                  }
                  x(1)+=0.12155306948221453*(1 + (-1.4 + x(1))*(-1.4 + x(1)))/(krmax-1);
              }

          }
      }}
  cout<<"Saving...";
  samples->save(filename);
  cout<<"Saved";
  } else
  {cout<<"loading...";  samples=new DataTable(filename); cout<<"Loaded";}; ///create new datatable from the file

  cout<<"Grid Complete: "<<samples->isGridComplete() <<" and ";
  cout<<samples->getNumVariables()<<" variables "<< samples->getNumSamples()<<" samples";
  char anykey;
  cin>>anykey;

  BSplineApproximant *bspline3;
  bspline3 = new BSplineApproximant(*samples, BSplineType::LINEAR);

//  HERE it hangs forever

**Comment by slizovskiy** [comment]
I will try the same code with much lower sample number now ( 4^6 samples)  and report on the result

**Comment by gablank** [comment]
Hello, Sergey.

Are you sure you are compiling the library in release mode? In the CMake step you need to use
cmake .. -DCMAKE_BUILD_TYPE=release
to get a release build, else it defaults to a debug build.

If the code outputs "Computing B-spline control points using ****** solver." then you are indeed using a debug build. In that case you need to delete CMakeCache.txt and rerun the CMake step with the additional option I specified.

**Comment by slizovskiy** [comment]
Hi!  Thank you for the prompt reply!  I am compiling in the release mode and do not get a message "Computing B-spline control points using ****** solver.".   With 6000 samples it takes around 20 seconds which would be ages with million samples.  I cannot understand  why should it take so long.

**Comment by gablank** [comment]
I agree that it shouldn't take that long with that few samples, but I am not familiar enough with the inner workings of the BSpline construction (that is @bgrimstad's work), but as he has written in the readme: "On a modern desktop computer the practical limit on the number of samples is about 100 000 when constructing a B-spline.". It seems you may have encountered a bug, but I am unable to test it right now, and @bgrimstad is probably asleep, so we can't give you a definite answer right now.

**Comment by slizovskiy** [comment]
Thanks!  There is one more issue to improve:  the spline is evaluated only inside the data domain.  This is a bit too restrictive as spline is very useful for extrapolation (at list a little bit) outside the data domain.

**Comment by slizovskiy** [comment]
I will have a look at different spline implementations e.g. here:   http://web.utk.edu/~asteine1/o2scl/html/tintp_section.html 
hope it will work faster.

**Comment by bgrimstad** [comment]
Hello Sergey,

You write:
> Hi! I have 6 variables on a regular grid, and constructing of even the LINEAR spline loads the processor forever. Is that a bug?

You have discovered what is known as "the curse of dimensionality" - the exponential growth with the number of dimensions. With 6-7 variables, even a low resolution grid will require a large amount of sample points. Consider 6 variables discretized with 10 values, then you reach 10^6 = 1 000 000 sample points. It is possible to build a B-spline from a sample of scattered points, however, it is then difficult to ensure sufficient coverage of the domain (the domain should not have any large holes without sample points). *This is why I consider the practical limit for multivariate function approximation with B-splines to be 6-7 variables.* 

Side note: the implementation in SPLINTER utilizes the sparse structure of B-splines. From my experience the bottleneck is not memory usage (a million sample points require less than 8 MB), but CPU time required to solve the sparse linear system that gives the B-spline coefficients (construction of the B-spline).

You write:
>The same code with much lower sample number now ( 4^6 samples) seems to work.... With 4096 sample points it takes around 10 seconds to construct the linear interpolation. Why is it so slow???

I do not know the specs of your computer, but that did seem a tad (not unreasonably) slow. I would have guessed around 2-3 seconds on a fast laptop.

A couple of suggestions that may help speed up the construction:
1. Try a different sparse linear solver in BSplineApproximant::computeCoefficients. See available solvers in linearsolvers.h.
2. Limit the number of variables and sample points
3. Use a more sparse sample (support for scattered sample points will be supported in the next  version of SPLINTER, but that does not help you now, unfortunately)
4. Create a CUDA interface for us:-)

You write:
>There is one more issue to improve: the spline is evaluated only inside the data domain. This is a bit too restrictive as spline is very useful for extrapolation (at list a little bit) outside the data domain.

This is not correct. With (p+1)-regular knot vectors, the B-spline has support only in the closure of the domain spanned by the sample points (the "data domain"). Furthermore, if you were to construct the B-spline without (p+1)-regular knot vectors to have support (a little bit) outside of the "data domain", the behavior of the spline in this outer region would be a question of design. For example, the B-spline is designed to have constant second-derivatives in this region. I suggest that you use linear regression or similar techniques for extrapolation purposes.

You write:
>I will have a look at different spline implementations e.g. here: http://web.utk.edu/~asteine1/o2scl/html/tintp_section.html hope it will work faster.

Please share your learnings with us. We are always interested in improving our code so that it may help others looking to perform multivariate function approximation.

Best of luck on your adventures!

Bjarne

**Comment by slizovskiy** [comment]
Dear Bjarne,    Thank you for your reply!   Probably I do not understand the algorithm used,  but the simplest one for e.g. linear spline would be to find the 2^d nearest points to a given one  and make a linear interpolation.   2^6  is not a terribly large number, so, it must be fast.   With higher degree N spline, it would probably be  (N+1)^d, which may indeed become large with growing dimension.

**Comment by bgrimstad** [comment]
Regarding a "linear" spline, the algorithm you sketch resembles what SPLINTER does. The linear system that is solved in BSplineApproximant::computeCoefficients gives the coefficients that ensures interpolation. The main difference to your algorithm is that SPLINTER does all the interpolations in one solve.

Please note that linear interpolation on simplices is *not* the same as linear spline interpolation on hyper-rectangles. Confusion often occurs because the term "linear" is abused: a "linear" spline in two variables is actually bi-linear and will contain products x1*x2. A "linear" spline in three variables is tri-linear and will contain products x1*x2*x3. And so on. 

I suggest that you measure the CPU time used by BSplineApproximant::computeCoefficients. If most of the time is used there, then you have your answer: it does take some time to perform all of the interpolations.

**Comment by slizovskiy** [comment]
ok, I see!   So, it computes all the interpolation at ones and will then work very fast at evaluate, right?   So, if I save the computed spline,  it will be easy to load and use it later.

**Comment by bgrimstad** [comment]
Exactly.

**Comment by slizovskiy** [comment]
I hope the time to compute the spline scales linearly with the number of samples.... Otherwise  it's hopeless for me...

**Comment by slizovskiy** [comment]
Seems that scaling is above linear in my case...  I am waiting for 4 hours with 200 000 samples in 6D   and still no result...   I will have to go for some simpler method.  It is easier for me to generate more points than use good-quality interpolation.

**Comment by gablank** [comment]
Sergey,

The runtime for building the BSpline should be Ω(n^2) and O(n^3), so for a large amount of samples the runtime kind of explodes. How often do you need to build the BSpline? How long can you afford to wait? If you have built the BSpline you can save it for later use.

**Comment by gablank** [comment]
Hello again, Sergey.

I have just tested the build time on my own (fairly slow) computer. It has an AMD Phenom II 1055T processor, which has rather bad single thread performance. I created a new branch with the code I used, called slizovskiy_performance_regression. You can pull it and test it for yourself to compare.

Building it with `cmake .. -DCMAKE_BUILD_TYPE=release` and running with
`time ./splinter-test`
results in this output:
```
Adding samples
Building linear BSpline from 4096 samples
Finished building BSpline

real	0m3.369s
user	0m3.056s
sys	0m0.308s
```

Note that the time usage includes the time taken to add the samples (which is negligible).

As you can see I get significantly better runtime than you, although when n grows it will still get intractable fast.

Note: I don't know if the function I sampled has any impact on the time taken to solve the linear system or not, but my time seems to agree with what @bgrimstad guessed, so I think the timing should be good.

**Comment by slizovskiy** [comment]
Hi!   When I have 1000 samples, I use Mathematica. I go to C++  when I really need  speed.  Are you sure that your algorithm cannot be improved to scale linearly with the number of samples?    It looks to me that  spline is a local thing and should be independent of the number of samples that are far-away. Thus, it must scale extensively with the number of samples.

**Comment by bgrimstad** [comment]
Sergey,

An interpolating spline can be constructed using various methods. These methods are often classified as being *local* or *global*. The former perform interpolation by only utilizing data local to the interpolation point, similar to what you suggest. These methods are fast, but often require additional information such as derivatives. The global methods, on the other hand, utilize all information when performing the interpolation. This is what SPLINTER does today and why a single LSE is solved to find the coefficients. A disadvantage by doing this is that solving the (possibly) large, but sparse LSE can be time consuming. An advantage is that derivatives are not required, only the sample points.

In short: your intuition is correct, but unfortunately SPLINTER does not currently have a local interpolation method.

You can read more about these matters in Lyche and Moerken's book *Spline methods draft (2011)*. It is freely available on the internet. Chap. 5.4 describes the interpolation method used by SPLINTER.

Best regards,

Bjarne

**Comment by slizovskiy** [comment]
Bjarne, 
 Thank you!  I got it.   My only suggestion is to write what you have written  with a larger font below the advertisement ;)))
   I will try simpler local algorithms now.
Best wishes,
Sergey
